File: paimon-core/src/main/java/org/apache/paimon/manifest/FileEntry.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.paimon.manifest;

import org.apache.paimon.data.BinaryRow;
import org.apache.paimon.data.InternalRow;
import org.apache.paimon.utils.FileStorePathFactory;
import org.apache.paimon.utils.Filter;
import org.apache.paimon.utils.Preconditions;

import javax.annotation.Nullable;

import java.util.Arrays;
import java.util.Collection;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;
import java.util.stream.Collectors;

import static org.apache.paimon.utils.ManifestReadThreadPool.randomlyExecuteSequentialReturn;
import static org.apache.paimon.utils.ManifestReadThreadPool.sequentialBatchedExecute;

/** Entry representing a file. */
public interface FileEntry {

    FileKind kind();

    BinaryRow partition();

    int bucket();

    int level();

    String fileName();

    Identifier identifier();

    String externalPath();

    BinaryRow minKey();

    BinaryRow maxKey();

    List<String> extraFiles();

    /**
     * The same {@link Identifier} indicates that the {@link ManifestEntry} refers to the same data
     * file.
     */
    class Identifier {
        public final BinaryRow partition;
        public final int bucket;
        public final int level;
        public final String fileName;
        public final List<String> extraFiles;
        @Nullable private final byte[] embeddedIndex;

        /* Cache the hash code for the string */
        private Integer hash;

        public Identifier(
                BinaryRow partition,
                int bucket,
                int level,
                String fileName,
                List<String> extraFiles,
                @Nullable byte[] embeddedIndex) {
            this.partition = partition;
            this.bucket = bucket;
            this.level = level;
            this.fileName = fileName;
            this.extraFiles = extraFiles;
            this.embeddedIndex = embeddedIndex;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) {
                return true;
            }
            if (o == null || getClass() != o.getClass()) {
                return false;
            }
            Identifier that = (Identifier) o;
            return bucket == that.bucket
                    && level == that.level
                    && Objects.equals(partition, that.partition)
                    && Objects.equals(fileName, that.fileName)
                    && Objects.equals(extraFiles, that.extraFiles)
                    && Objects.deepEquals(embeddedIndex, that.embeddedIndex);
        }

        @Override
        public int hashCode() {
            if (hash == null) {
                hash =
                        Objects.hash(
                                partition,
                                bucket,
                                level,
                                fileName,
                                extraFiles,
                                Arrays.hashCode(embeddedIndex));
            }
            return hash;
        }

        @Override
        public String toString() {
            return "{partition="
                    + partition
                    + ", bucket="
                    + bucket
                    + ", level="
                    + level
                    + ", fileName="
                    + fileName
                    + ", extraFiles="
                    + extraFiles
                    + ", embeddedIndex="
                    + Arrays.toString(embeddedIndex)
                    + '}';
        }

        public String toString(FileStorePathFactory pathFactory) {
            return pathFactory.getPartitionString(partition)
                    + ", bucket "
                    + bucket
                    + ", level "
                    + level
                    + ", file "
                    + fileName
                    + ", extraFiles "
                    + extraFiles
                    + ", embeddedIndex "
                    + Arrays.toString(embeddedIndex);
        }
    }

    static <T extends FileEntry> Collection<T> mergeEntries(Iterable<T> entries) {
        LinkedHashMap<Identifier, T> map = new LinkedHashMap<>();
        mergeEntries(entries, map);
        return map.values();
    }

    static void mergeEntries(
            ManifestFile manifestFile,
            List<ManifestFileMeta> manifestFiles,
            Map<Identifier, ManifestEntry> map,
            @Nullable Integer manifestReadParallelism) {
        mergeEntries(
                readManifestEntries(manifestFile, manifestFiles, manifestReadParallelism), map);
    }

    static <T extends FileEntry> void mergeEntries(Iterable<T> entries, Map<Identifier, T> map) {
        for (T entry : entries) {
            Identifier identifier = entry.identifier();
            switch (entry.kind()) {
                case ADD:
                    Preconditions.checkState(
                            !map.containsKey(identifier),
                            "Trying to add file %s which is already added.",
                            identifier);
                    map.put(identifier, entry);
                    break;
                case DELETE:
                    // each dataFile will only be added once and deleted once,
                    // if we know that it is added before then both add and delete entry can be
                    // removed because there won't be further operations on this file,
                    // otherwise we have to keep the delete entry because the add entry must be
                    // in the previous manifest files
                    if (map.containsKey(identifier)) {
                        map.remove(identifier);
                    } else {
                        map.put(identifier, entry);
                    }
                    break;
                default:
                    throw new UnsupportedOperationException(
                            "Unknown value kind " + entry.kind().name());
            }
        }
    }

    static Iterable<ManifestEntry> readManifestEntries(
            ManifestFile manifestFile,
            List<ManifestFileMeta> manifestFiles,
            @Nullable Integer manifestReadParallelism) {
        return sequentialBatchedExecute(
                file -> manifestFile.read(file.fileName(), file.fileSize()),
                manifestFiles,
                manifestReadParallelism);
    }

    static Set<Identifier> readDeletedEntries(
            ManifestFile manifestFile,
            List<ManifestFileMeta> manifestFiles,
            @Nullable Integer manifestReadParallelism) {
        return readDeletedEntries(
                m ->
                        manifestFile.read(
                                m.fileName(),
                                m.fileSize(),
                                Filter.alwaysTrue(),
                                deletedFilter(),
                                Filter.alwaysTrue()),
                manifestFiles,
                manifestReadParallelism);
    }

    static <T extends FileEntry> Set<Identifier> readDeletedEntries(
            Function<ManifestFileMeta, List<T>> manifestReader,
            List<ManifestFileMeta> manifestFiles,
            @Nullable Integer manifestReadParallelism) {
        manifestFiles =
                manifestFiles.stream()
                        .filter(file -> file.numDeletedFiles() > 0)
                        .collect(Collectors.toList());
        Function<ManifestFileMeta, List<Identifier>> processor =
                file ->
                        manifestReader.apply(file).stream()
                                // filter again, ensure is delete
                                .filter(e -> e.kind() == FileKind.DELETE)
                                .map(FileEntry::identifier)
                                .collect(Collectors.toList());
        Iterator<Identifier> identifiers =
                randomlyExecuteSequentialReturn(processor, manifestFiles, manifestReadParallelism);
        Set<Identifier> result = ConcurrentHashMap.newKeySet();
        while (identifiers.hasNext()) {
            result.add(identifiers.next());
        }
        return result;
    }

    static Filter<InternalRow> deletedFilter() {
        Function<InternalRow, FileKind> getter = ManifestEntrySerializer.kindGetter();
        return row -> getter.apply(row) == FileKind.DELETE;
    }

    static Filter<InternalRow> addFilter() {
        Function<InternalRow, FileKind> getter = ManifestEntrySerializer.kindGetter();
        return row -> getter.apply(row) == FileKind.ADD;
    }
}

File: paimon-python/pypaimon/manifest/schema/file_entry.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
################################################################################

"""Entry representing a file.

Follows the design of Java's org.apache.paimon.manifest.FileEntry.
"""


class FileEntry:
    """Entry representing a file.

    The same Identifier indicates that the FileEntry refers to the same data file.
    """

    class Identifier:
        """Unique identifier for a file entry.

        Uses partition, bucket, level, fileName, extraFiles,
        embeddedIndex and externalPath to identify a file.
        """

        def __init__(self, partition, bucket, level, file_name,
                     extra_files, embedded_index, external_path):
            self.partition = partition
            self.bucket = bucket
            self.level = level
            self.file_name = file_name
            self.extra_files = extra_files
            self.embedded_index = embedded_index
            self.external_path = external_path
            self._hash = None

        def __eq__(self, other):
            if self is other:
                return True
            if other is None or not isinstance(other, FileEntry.Identifier):
                return False
            return (self.bucket == other.bucket
                    and self.level == other.level
                    and self.partition == other.partition
                    and self.file_name == other.file_name
                    and self.extra_files == other.extra_files
                    and self.embedded_index == other.embedded_index
                    and self.external_path == other.external_path)

        def __hash__(self):
            if self._hash is None:
                self._hash = hash((
                    self.partition,
                    self.bucket,
                    self.level,
                    self.file_name,
                    self.extra_files,
                    self.embedded_index,
                    self.external_path,
                ))
            return self._hash

    def identifier(self):
        """Build a unique Identifier for this file entry.

        Returns:
            An Identifier instance.
        """
        extra_files = (tuple(self.file.extra_files)
                       if self.file.extra_files else ())
        return FileEntry.Identifier(
            partition=self.partition,
            bucket=self.bucket,
            level=self.file.level,
            file_name=self.file.file_name,
            extra_files=extra_files,
            embedded_index=self.file.embedded_index,
            external_path=self.file.external_path,
        )

    @staticmethod
    def merge_entries(entries):
        """Merge file entries: ADD and DELETE of the same file cancel each other.

        - ADD: if identifier already in map, raise error; otherwise add to map.
        - DELETE: if identifier already in map, remove both (cancel);
          otherwise add to map.

        Args:
            entries: Iterable of FileEntry.

        Returns:
            List of merged FileEntry values, preserving insertion order.

        Raises:
            RuntimeError: If trying to add a file that is already in the map.
        """
        entry_map = {}

        for entry in entries:
            entry_identifier = entry.identifier()
            if entry.kind == 0:  # ADD
                if entry_identifier in entry_map:
                    raise RuntimeError(
                        "Trying to add file {} which is already added.".format(
                            entry.file.file_name))
                entry_map[entry_identifier] = entry
            elif entry.kind == 1:  # DELETE
                if entry_identifier in entry_map:
                    del entry_map[entry_identifier]
                else:
                    entry_map[entry_identifier] = entry
            else:
                raise RuntimeError(
                    "Unknown entry kind: {}".format(entry.kind))

        return list(entry_map.values())
